(test) Authentication Tests

klinkoberstar opened the pull request:
This pull request creates a series of tests for the authorization routes and provides some security tests for authentication.

This PR will close #126 and close #127

## Describe your changes

Updated a few things in the conftest.py file:
- We need a secret key for the in-memory DB to run, added a placeholder key for testing
- Added 3 fixtures:
  - capture_templates: extract the template of the current page, so we can assert that the proper template is rendered, after a successful 200 response
  - test_user: creates an instance of the test user in the database. This is needed to test the login functionality for a user already in the database, as well as to test the password changes
  - login_client: creates a logged-in state, to test functionalities that require the user already be logged in (i.e. change password, etc.)

Three different test files: 
- auth_basic_test.py: tests to check login, logout, and sign-up functionality
- auth_change_password_test.py: a suite of tests specifically focused on different scenarios under which the password is changed
- auth_security_test.py: additional tests to put pressure on authentication security

## Non-obvious technical information


## Checklist before requesting a review
- [X] pre-commit run --all-files (run before pushing)
- [X] pytest if applicable
- [X] Link issue
- [X] Update relevant documentation if applicable: doc strings, readme, poetry.

aaronhaefner commented:
@klinkoberstar can you merge in main and address merge conflicts? We just merged #129 which adds some rules to docstrings.

klinkoberstar commented:
> @klinkoberstar can you merge in main and address merge conflicts? We just merged #129 which adds some rules to docstrings.

I've merged in main, addressed the conflicts, and updated the docstrings to comply with new reqs!

aaronhaefner commented on tests/conftest.py:
```diff
@@ -114,3 +120,46 @@ def create_logger(name: str, level=logging.INFO) -> logging.Logger:
+    user_password = generate_password_hash(
+        "TestP@ssword!", method="pbkdf2:sha256"
+    )
+    test_user = User(email="test@example.com", password=user_password)
+    db.session.add(test_user)
+    db.session.commit()
+
+    yield test_user
+
+    db.session.delete(test_user)
```
Use a different name than `test_user` again since its the fixture name too.

aaronhaefner commented:
I would also add some logging to your tests throughout. `db_test.py` has examples.

aaronhaefner commented on tests/auth_change_password_test.py:
```diff
@@ -0,0 +1,224 @@
+        test_user: User instance for which the test is run.
+        login_client: Session object for the logged-in client.
+    """
+    response = client.get("/change_password", follow_redirects=True)
```
should this be `login_client`?

klinkoberstar replied:
The way I have it set up, the login_client just logs the test user in at the start of the test, it's not a separate logged-in instance of the app so it would still be client.get

realizing the name is confusing - in my head it was like  login_client = "log into the client" which is not at all how it reads here. i can change the naming to make it less confusing!

aaronhaefner replied:
I see that makes sense! You can keep the naming, just add explanation to `readme` for the new fixtures.

klinkoberstar replied:
ended up changing the name bc i think it is confusing - updated the readme as well!

aaronhaefner commented on tests/auth_basic_test.py:
```diff
@@ -0,0 +1,189 @@
+    response = client.get("/logout", follow_redirects=True)
+    assert response.status_code == 200
+    assert len(capture_templates) == 2
+    assert capture_templates[1][0].name == "home.html", "Wrong template used"
```
Could you add in a check here that tries to access a page that is `@login_required` after you logout? Fine if you use/edit this.

```
profile_response = client.get("/profile", follow_redirects=True)
assert profile_response.status_code == 200, "Response after redirection"
assert b"Login" in profile_response.data, "User prompted to log in"
assert "login.html" in [
    template[0].name for template in capture_templates
], "Redirect to login page after trying to access profile post-logout"
```

klinkoberstar commented:
> I would also add some logging to your tests throughout. `db_test.py` has examples.

logging has been added!
